### DISCUSSION TITLE
Prevent generating `maybe_`-prefixed method, and a way to provide the default value from within `fn new`

### DISCUSSION THREAD

**Discussion by teohhanhui** [elastio/bon#203] [Q&A]
1. I think there should be a way to prevent generating the `maybe_`-prefixed method, i.e. only make the call optional with no way of explicitly setting `None`.

2. It'd be nice to have a way to provide the default value from inside the function body of `fn new`, e.g.

```rust
struct TilePos {
    x: u32,
    y: u32,
}

#[bon]
impl TilePos {
    #[builder]
    fn new(x: u32, maybe_y: Option<u32>) -> Self {
        Self {
            x,
            y: y.unwrap_or_else(|| x * 2),
        }
    }
}
```

**Comment by Veetaha** [comment]
Hi! 

(1) Could you elaborate so I better understand the motivation? Why would you like to remove the `maybe_`-prefixed method? In what specific use case do you think it would be excessive?

(2) It's already possible to provide the default value from inside the function body

```rust
use bon::bon;

struct TilePos {
    x: u32,
    y: u32,
}

#[bon]
impl TilePos {
    #[builder]
    fn new(x: u32, y: Option<u32>) -> Self {
        Self {
            x,
            y: y.unwrap_or_else(|| x * 2),
        }
    }
}

fn main() {
    // You can get the default value if you don't call `y/maybe_y` setters at all
    let value = TilePos::builder().x(3).build();
    assert_eq!(value.y, 6);

    // ...or if you pass `None` to `maybe_y` setter
    let value = TilePos::builder().x(3).maybe_y(None).build();
    assert_eq!(value.y, 6);
}
```

Alternatively, the same could be written with `derive(Builder)` syntax because you can access values of other members declared higher in the struct inside of the `#[builder(default = ...)]`. See [Evaluation Context](https://bon-rs.com/reference/builder/member/default#evaluation-context) paragraph in `#[builder(default)]` API reference.
 
 ```rust
 #[derive(bon::Builder)]
struct TilePos {
    x: u32,

    #[builder(default = x * 2)]
    y: u32,
}

fn main() {
    // You can get the default value if you don't call `y/maybe_y` setters at all
    let value = TilePos::builder().x(3).build();
    assert_eq!(value.y, 6);

    // ...or if you pass `None` to `maybe_y` setter
    let value = TilePos::builder().x(3).maybe_y(None).build();
    assert_eq!(value.y, 6);
}
```

**Reply by teohhanhui** [reply]
> (1) Could you elaborate so I better understand the motivation? Why would you like to remove the maybe_-prefixed method? In what specific use case do you think it would be excessive?

It's just that I would not want it to be available in some cases. Of course, it's often a matter of personal preference in terms of what kind of API we want to expose... The API surface is important for compatibility when switching between libraries, for example.

My initial impression of this library is that it seems to follow a similar philosophy as `thiserror`, where you can transparently switch to a manual implementation without changing the API. Is that right?

> (2) It's already possible to provide the default value from inside the function body

I think it could be better illustrated in the docs, perhaps. I mostly only managed to find `#[builder(default = ...)] `

**Reply by Veetaha** [reply]
(1) To be clear, there is a good motivation behind generating `maybe_` setters to not even allow the user to disable it. It's not obvious, but let me explain.

If you don't have a `maybe_` setter it becomes very annoying in a situation when you need to dynamically decide whether to use `None`/default value or not, e.g. when you already have a variable of type `Option<...>`:


```rust
#[derive(bon::Builder)]
struct Example {
    x: Option<u32>,
}

fn main() {
    // Could be any value (decided at runtime via user input, for example)
    let value: Option<u32> = None;

    let example = if let Some(value) = value {
        Example::builder()
            .x(value)
            .build()
    } else {
        Example::builder().build()
    };
}
```

Here you have to write this big `if` that conditionally calls the `x(...)` setter. You can't avoid that. Try playing with this yourself and you'll understand why I think this is so annoying. This is actually one of the reasons I created `bon`, beause `typed-builder` doesn't genereate `maybe_` setters by default, which makes using such API inconvenient.

**Reply by teohhanhui** [reply]
That's perfectly understandable as a default, and I think it's a good default. But personally I find the `maybe_`-prefixed method redundant, and somewhat confusing (especially in terms of naming).

p/s: A serious offender of this `maybe_`-abuse is WordPress, from what I remember.

**Reply by Veetaha** [reply]
(2)
>  I think it could be better illustrated in the docs, perhaps. I mostly only managed to find #[builder(default = ...)] 

I see, I'll add this pattern to the guide then

**Reply by Veetaha** [reply]
> That's perfectly understandable as a default, and I think it's a good default. But personally I find the maybe_-prefixed method redundant, and somewhat confusing (especially in terms of naming).

You can tweak the name of that setter with [`#[builder(setters(option_fn = some_other_name))]`](https://bon-rs.com/reference/builder/member/setters).

Given that you have an ability to rename that setter to give it a better name do you still think that removing it is a good idea? Given the annoyance of dynamic `Option<T>` value setting that I pointed out higher? I believe the main critic you have is the name of the `maybe_` setter, not the fact of its presence, right?

**Reply by Veetaha** [reply]
> p/s: A serious offender of this maybe_-abuse is WordPress, from what I remember.

Never used WordPress, but I guess you had bad experience with some API in it that uses `maybe_` in its name. However, I believe the logic in `bon` API is completely different and has good motivation behind it

**Reply by teohhanhui** [reply]
> I believe the main critic you have is the name of the maybe_ setter, not the fact of its presence, right?

Unfortunately I don't think there's a good name for that. And it'd break the reading flow of the fluent API anyway...

But I sympathize with the rationale in having it provided by default.

That said, your example would probably usually be like this instead:

```rust
#[derive(bon::Builder)]
struct Example {
    x: Option<u32>,
}

fn main() {
    // Could be any value (decided at runtime via user input, for example)
    let value: Option<u32> = None;

    let builder = Example::builder();
    // Do some shared things...
    let builder = if let Some(value) = value {
        builder.x(value)
    } else {
        builder
    };
    // Do some more shared things...
    let example = builder.build();
}
```

**Reply by Veetaha** [reply]
Nope, unfortunately, the example you shown above doesn't compile with the following error: 

```
`if` and `else` have incompatible types
expected struct `ExampleBuilder<example_builder::SetX>`
   found struct `ExampleBuilder<example_builder::Empty>`
```

This is because the builder uses a typestate to check for overwrites. It changes the type when you call a setter to make sure you don't accidentally call the setter for the same member twice.

**Reply by Veetaha** [reply]
You could theoretically disable that check with [`#[builder(overwritable)]`](https://bon-rs.com/reference/builder/member/overwritable), in which case the example above would start compiling. Would you give up on overwrite checks in your use case to make the code shown above compile?

**Reply by Veetaha** [reply]
To be clear, I need a convincing use case example, where disabling the `option_fn` (i.e. `maybe_` setter) obviously improves the API. You mentioned that there isn't a good name for this setter in your case and that it breaks the fluent API.

Could you show the exact code you mean in this case? I need to have that context as well, otherwise I'm not convinced. You can omit impl details or come up with the alternative example if the code in question is under NDA

**Reply by teohhanhui** [reply]
```
`if` and `else` have incompatible types
expected struct `ExampleBuilder<example_builder::SetX>`
   found struct `ExampleBuilder<example_builder::Empty>`
```

Uhh, that's unfortunate but understandable, given the whole point of the typestate pattern being used...

> To be clear, I need a convincing use case example

I don't have such an example. Sorry... Just thinking ahead in terms of the API being exposed by the builder, that's all.

Especially given my impression here:

> My initial impression of this library is that it seems to follow a similar philosophy as thiserror, where you can transparently switch to a manual implementation without changing the API. Is that right?

**Reply by Veetaha** [reply]
> My initial impression of this library is that it seems to follow a similar philosophy as thiserror, where you can transparently switch to a manual implementation without changing the API. Is that right?

Yes, your impression is correct. The generated builder's API doesn't couple it with `bon` specifically, and you could theoretically replace it with a manual impl.

I'll keep this at the current state where it's not possible to disable the `maybe_` setter, until I see a compelling use case for disabling it. This is to make sure users don't shoot themselves in the foot with the dynamic `Option<T>` handling annoyances and thinking that they could use an `if {}` as you shown above to workaround this (which has own nuances).

Note that it is physically possible to hide that setter anyway by making the setters private `#[builder(setters(vis = ""))]`, and write a [custom setter](https://bon-rs.com/guide/typestate-api/custom-methods#custom-setters). It's just that I don't think anyone would ever want to do this at this point. I'm ready to be proven wrong though
